sync-core/src/test/java/com/cloudant/sync/query/AbstractQueryExecutorTestBase.java

```diff
@@ -0,0 +1,280 @@
+//  Copyright (c) 2014 Cloudant. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+
+package com.cloudant.sync.query;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+import com.cloudant.sync.datastore.DocumentBodyFactory;
+import com.cloudant.sync.datastore.DocumentRevision;
+import com.cloudant.sync.datastore.MutableDocumentRevision;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *  The aim is to make sure that the post hoc matcher class behaves the
+ *  same as the SQL query engine.
+ *
+ *  To accomplish this, we test the entire test execution pipeline contained
+ *  within this class using {@link com.cloudant.sync.query.QueryExecutorTest}
+ *  which exercises the SQL query engine matching functionality.  We then test
+ *  the same test execution pipeline contained within this class using
+ *  {@link com.cloudant.sync.query.QueryExecutorUnindexedTest} which in turn
+ *  exercises the post hoc matcher matching functionality.
+ */
+public abstract class AbstractQueryExecutorTestBase extends AbstractIndexTestBase {
+
+    @Override
+    public void setUp() throws SQLException {
```

Comment:
I don't think this works correctly, because you appear to create the same class of IndexManager object (in `AbstractIndexTestBase` `setUp()`) for all of the query test classes, which will use both SQL indexes and the unindexed matcher?

You need sub-classes of IndexManager specific to testing, which override:

* the `QueryTranslator#translateQuery()` to always return a query which matches all document (so only the unindexed matcher is being tested).
* the `QueryTranslator#matcherForIndexCoverage()` to always return `null` (so only the SQL query engine is tested).